frame/name-service/src/lib.rs

```diff
@@ -0,0 +1,487 @@
+		CommitmentNotFound,
+		/// This registration does not exist.
+		RegistrationNotFound,
+		/// The sender is not the registration owner.
+		NotRegistrationOwner,
+		/// This registration has expired.
+		RegistrationExpired,
+		/// This registration has not yet expired.
+		RegistrationNotExpired,
+	}
+
+	// Your Pallet's callable functions.
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		// TODO: Should we allow registration on behalf of?
+		#[pallet::weight(0)]
+		pub fn commit(
+			origin: OriginFor<T>,
+			who: T::AccountId,
+			commitment_hash: CommitmentHash,
+		) -> DispatchResult {
+			let sender = ensure_signed(origin)?;
+
+			ensure!(!Commitments::<T>::contains_key(commitment_hash), Error::<T>::AlreadyCommitted);
+
+			let block_number = frame_system::Pallet::<T>::block_number();
+			let deposit = T::CommitmentDeposit::get();
+
+			T::Currency::reserve(&sender, deposit)?;
+
+			let commitment = Commitment { who: who.clone(), when: block_number, deposit };
+
+			Commitments::<T>::insert(commitment_hash, commitment);
+			Self::deposit_event(Event::<T>::Committed { sender, who, hash: commitment_hash });
+
+			Ok(())
+		}
+
+		#[pallet::weight(0)]
+		pub fn reveal(
+			origin: OriginFor<T>,
+			name: Vec<u8>,
+			secret: u64,
+			periods: u32,
+		) -> DispatchResult {
+			let sender = ensure_signed(origin)?;
+			let commitment_hash = sp_io::hashing::blake2_256(&(&name, secret).encode());
+
+			let commitment =
+				Commitments::<T>::get(commitment_hash).ok_or(Error::<T>::CommitmentNotFound)?;
+
+			let name_hash = sp_io::hashing::blake2_256(&name);
+			let block_number = frame_system::Pallet::<T>::block_number();
+
+			ensure!(
+				block_number > commitment.when.saturating_add(T::MinimumCommitmentPeriod::get()),
+				Error::<T>::TooEarlyToReveal
+			);
+
+			if Self::is_available(name_hash, block_number) {
+				let fee = Self::registration_fee(name.clone(), periods);
+
+				let imbalance = T::Currency::withdraw(
+					&sender,
+					fee,
+					WithdrawReasons::FEE,
+					ExistenceRequirement::KeepAlive,
+				)?;
+
+				T::RegistrationFeeHandler::on_unbalanced(imbalance);
+
+				let deposit = T::NameDeposit::get();
+				Self::do_register(name_hash, commitment.who.clone(), deposit, periods)?;
+			}
+
+			T::Currency::unreserve(&commitment.who, commitment.deposit);
+			Commitments::<T>::remove(commitment_hash);
+
+			Ok(())
+		}
+
+		#[pallet::weight(0)]
+		pub fn transfer(
+			origin: OriginFor<T>,
+			to: T::AccountId,
+			name_hash: NameHash,
+		) -> DispatchResult {
+			let sender = ensure_signed(origin)?;
+			let block_number = frame_system::Pallet::<T>::block_number();
+
+			Registrations::<T>::try_mutate(name_hash, |maybe_registration| {
+				let r = maybe_registration.as_mut().ok_or(Error::<T>::RegistrationNotFound)?;
+				ensure!(r.owner == sender, Error::<T>::NotRegistrationOwner);
+				ensure!(r.expiry > block_number, Error::<T>::RegistrationExpired);
+
+				r.owner = to.clone();
+
+				// TODO: transfer deposit over?
```

Comment:
Still need to pay back some deposit?

Reply:
Deposits are currently made for subnodes (with `repatriate_reserved` when a new owner is set), but currently not for top level registrations, reasoning being that registration + length fees should be enough to deterrent spamming and hoarding to some extent. Saying that, I think @shawntabrizi is interested in exploring a reference counted deposit here.